h/views/api/orcid.py

```diff
@@ -0,0 +1,167 @@
+@view_defaults(request_method="GET", route_name="orcid.oauth.authorize")
+class AuthorizeViews:
+    def __init__(self, request: Request) -> None:
+        self._request = request
+
+    @view_config(is_authenticated=True)
+    def authorize(self):
+        host = self._request.registry.settings["orcid_host"]
+        client_id = self._request.registry.settings["orcid_client_id"]
+        state = RetrieveOAuthCallbackSchema(self._request).state_param()
+
+        params = {
+            "client_id": client_id,
+            "response_type": "code",
+            "redirect_uri": self._request.route_url("orcid.oauth.callback"),
+            "state": state,
+            "scope": "openid",
+        }
+        return HTTPFound(
+            location=urlunparse(
+                (
+                    "https",
+                    host,
+                    "oauth/authorize",
+                    "",
+                    urlencode(params),
+                    "",
+                )
+            )
+        )
+
+    @notfound_view_config()
+    def notfound(self):
+        return HTTPUnauthorized()
+
+
+class AccessDeniedError(Exception):
+    pass
+
+
+class UserConflictError(Exception):
+    pass
+
+
+@view_defaults(request_method="GET", route_name="orcid.oauth.callback")
+class CallbackViews:
+    def __init__(self, request: Request) -> None:
+        self._request = request
+        self._orcid_client = request.find_service(ORCIDClientService)
+        self._user_service = request.find_service(name="user")
+
+    @view_config(is_authenticated=True)
+    def callback(self):
+        try:
+            callback_data = RetrieveOAuthCallbackSchema(self._request).validate(
+                dict(self._request.params)
+            )
+        except ValidationError as err:
+            if self._request.params.get("error") == "access_denied":
+                # The user clicked the deny button on ORCID's page.
+                raise AccessDeniedError from err
+
+            # We received an invalid or unexpected redirect from ORCID.
+            raise
+
+        orcid = self._orcid_client.get_orcid(callback_data["code"])
+
+        already_connected_user = self._user_service.fetch_by_identity(
+            IdentityProvider.ORCID, orcid
+        )
+
+        if already_connected_user:
+            if already_connected_user == self._request.user:
+                # This ORCiD is already connected to this Hypothesis account,
+                # so nothing more needs to be done.
+                return HTTPFound(location=self._request.route_url("account"))
+
+            # Oops, this ORCiD is already connected to a *different* Hypothesis
+            # account.
+            raise UserConflictError
+
+        # This ORCiD isn't connected to a Hypothesis account yet.
+        # Let's go ahead and connect it to the user's account.
+        self._orcid_client.add_identity(self._request.user, orcid)
+        return HTTPFound(location=self._request.route_url("account"))
```

Comment:
I think we may want to add a flash message for the successful case as well. Maybe something like this:

> **ORCiD <ID> connected ✓**
> 
> You can now use your ORCiD to log in to Hypothesis.

If adding this message it should also be shown for the `already_connected_user == self._request.user` case above as well. The code can probably be refactored a bit to avoid duplication:

```python
if already_connected_user and already_connected_user != self._request.user:
    raise UserConflictError

if not already_connected_user:
    self._orcid_client.add_identity(self._request.user, orcid)

self._request.session.flash(<success_message>)
return HTTPFound(location=self._request.route_url("account"))
```